c7n/resources/ram.py

```diff
@@ -0,0 +1,248 @@
+            for assoc in [a for a in assocs if a['status'] == 'ASSOCIATED']:
+                associations_map.setdefault(assoc['resourceShareArn'], []).append(assoc)
+            for r in resources:
+                r.setdefault(
+                    self.associations_attribute, []
+                ).extend(
+                    associations_map.get(r['resourceShareArn'], [])
+                )
+        return resources
+
+    def process(self, resources, event=None):
+        results = []
+        resources = self.get_share_associations(resources)
+        for r in resources:
+            allowed_entities = set(self.manager.config.account_id)
+            allowed_entities = allowed_entities.union(self.data.get('allowlist_entities', ()))
+            if 'allowlist_entities_from' in self.data:
+                values = ValuesFrom(self.data['allowlist_entities_from'], self.manager)
+                allowed_entities = allowed_entities.union(values.get_values())
+            violations = [
+                assoc for assoc in r[self.associations_attribute]
+                if assoc['associatedEntity'] not in allowed_entities
+            ]
+            if violations:
+                r[self.annotation_key] = violations
+                results.append(r)
+        return results
+
+
+@ResourceShareSelf.action_registry.register('disassociate')
+class DisassociateResourceShare(Action):
+    """Action to disassociate principals from a Resource Share
+    https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ram/client/disassociate_resource_share.html
+    :example:
+
+    .. code-block:: yaml
+
+        policies:
+          - name: disassociate-ram-resource-share
+            resource: resource-share-self
+            filters:
+            - type: external-share
+              allowlist_entities:
+                - "123456789012"
+                - o-abcd1234
+            actions:
+            - type: disassociate
+              principals: matched
+    """
+
+    schema = type_schema(
+        'disassociate',
+        required=['principals'],
+        principals={'enum': ['matched', 'all']},
+    )
+    permissions = ('ram:DisassociateResourceShare',)
+
+    def validate(self):
+        if self.data.get('principals') == 'matched':
+            ftypes = {f.type for f in self.manager.iter_filters()}
+            if 'external-share' not in ftypes:
+                raise PolicyValidationError(
+                    "external-share filter is required when principals is 'matched'"
+                )
+        return self
+
+    def process(self, resources):
+        _all = self.data.get('principals') == 'all'
+
+        client = local_session(
+            self.manager.session_factory
+        ).client(self.manager.resource_type.service)
+
+        if _all and ExternalShareFilter.associations_attribute not in resources[0]:
+            share_arns = [r['resourceShareArn'] for r in resources]
```

Comment:
i do wonder about using composition and instantiating the external share filter here to unify the fetching of the associations, in order to dedupe the logic here (`get_share_associations`)